### PR TITLE
Extract teradata tasks

### PR DESCRIPTION
**Pull request by ishmum123**

### CODE REVIEW

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactory.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+  /**
+   * For each potential expression in EXPRESSIONS, work out whether this Teradata accepts it, and if
+   * so, use it as part of the eventual query.
+   *
+   * @param predicate A predicate to compute whether a given expression is legal.
+   * @return A SQL query containing every legal expression from EXPRESSIONS.
+   */
+  @Override
+  public String getSql(BiPredicate<String, String> predicate) {
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactory.java:61]: The list of available columns doesn't change during a single dumper invocation. The predicate is passed to the query building logic in the existing code, but it is a bit of a hack. I would suggest to decouple filtering of available columns from query building, so that this query factory can accept a list of available columns.

File: dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactoryBuilderTest.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  @Test
+  public void test_buildAssessmentFactory() throws IOException, MetadataDumperUsageException {
+    //  Arrange
+    ConnectorArguments arguments =
+        new ConnectorArguments(new String[] {"--connector", "teradata", "--assessment"});
+    ZonedDateTime endDate = ZonedDateTime.parse("2023-03-09T00:00:00Z");
+    ZonedDateTime startDate = endDate.minusDays(5).truncatedTo(ChronoUnit.DAYS);
+    ZonedInterval interval = new ZonedInterval(startDate, endDate);
+    String expectedQuery =
+        "SELECT L.QueryID, ST.SQLRowNo, ST.SQLTextInfo, L.AbortFlag, L.AcctString,"
```
Comment (dawidxc) [inline dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactoryBuilderTest.java:105]: I would suggest to put the test queries in a more readable form in a plain text `*.sql` file in the `src/test/resources` directory. One query per file.

Compare:
https://github.com/google/dwh-migration-tools/blob/858e7988f048024eeb9593adaf5d84a395c2c1e0/dumper/app/src/test/resources/dumper-test/oracle.sql

File: dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataLogsConnectorTest.java [hunk 3]
```diff
@@ -62,148 +67,174 @@
...
-        TeradataLogsConnector.TeradataLogsJdbcTask.EXPRESSION_VALIDITY_QUERY = "SELECT %s FROM %s FETCH FIRST 1 ROW ONLY";
-
-        MetadataDumper dumper = new MetadataDumper();
-        dumper.run(args.toArray(ArrayUtils.EMPTY_STRING_ARRAY));
-
-        // TODO: Use ZipValidator to assert that all N_QUERY_LOGS entries are present.
+  }
+
+  @Test
+  public void testConnectorSynthetic() throws Exception {
```
Comment (dawidxc) [inline dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataLogsConnectorTest.java:140]: I haven't checked exactly, but it looks like this test hasn't changed. If it's only formatting that was changed, I would suggest to move the formatting changes to a separate PR. Similarly for other cases, where only formatting was changed.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactoryBuilder.java [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+    // preventing that in the future. To do that, we should require getQueryLogEarliestTimestamp()
+    // to parse and return an ISO instant, not a database-server-specific format.
+    if (!StringUtils.isBlank(earliestTimestamp)) {
+      if (isLSql) {
+        conditions.add("ST.CollectTimeStamp >= " + earliestTimestamp);
+      } else {
+        conditions.add("L.StartTime >= " + earliestTimestamp);
+      }
+    }
+    if (!(isOldVersion || isAssessment)) conditions.add("L.UserName <> 'DBC'");
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactoryBuilder.java:249]: I would suggest to always use block (curly braces) for the `if` instruction.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactoryBuilder.java [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    addConditions(arguments.getQueryLogEarliestTimestamp());
+    addOrderColumns();
+    selectExpressions();
+  }
+
+  private void parseAlternates(List<String> alternates) throws MetadataDumperUsageException {
+    logTable = isAssessment ? ASSESSMENT_DEF_LOG_TABLE : DEF_LOG_TABLE;
+    queryTable = DEF_QUERY_TABLE;
+    if (alternates != null && !alternates.isEmpty()) {
+      if (alternates.size() != 2)
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactoryBuilder.java:228]: I would suggest to always use block (curly braces) for the `if` instruction.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactoryBuilder.java [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+  private ZonedInterval interval;
+  private String logTable;
+  private String queryTable;
+  private List<String> conditions;
+  private List<String> orderColumns;
+  private List<String> expressions;
+  private boolean isAssessment;
+  // For Terdata version <= 14
+  private boolean isOldVersion;
+  private final boolean isLSql;
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactoryBuilder.java:192]: Correct me if I'm wrong, but the 2 fields above (`isOldVersion` and `isLSql`) would not be needed, if the class `Teradata14LogsConnector.java` was reverted. I assume the behavior for Teradata<=14 would not change.

Advantages:
- less effort now
- `EXPRESSIONS_LSQL_TBL` and `EXPRESSIONS_LOG_TBL` constants and `enumNames` method could be removed from this class
- any code related to the 2 boolean fields (above) could also be simplified
- there would be no need to add tests for Teradata<=14 now

Also, if in the future it turns out that Teradata<=14 is needed, we will still be able to change the code of the old connector in a separate PR.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataLogsConnector.java [hunk 7]
```diff
@@ -17,433 +17,88 @@
...
+      String file =
+          ZIP_ENTRY_PREFIX
+              + DateTimeFormatter.ISO_OFFSET_DATE_TIME.format(interval.getStartUTC())
+              + ".csv";
+      SqlQueryFactory factory = builder.within(interval).build();
+      out.add(new TeradataLogsJdbcTask(file, state, factory).withHeaderClass(headerClass));
     }
+  }
 
-    protected static class TeradataLogsJdbcTask extends AbstractJdbcTask<Void> {
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataLogsConnector.java:146]: I would suggest to extract the `TeradataLogsJdbcTask` to a separate file in one PR and do other changes in separate PRs.

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataLogsJdbcTask.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+  @VisibleForTesting public static String EXPRESSION_VALIDITY_QUERY = "SELECT TOP 1 %s FROM %s";
+
+  private final SharedState state;
+  private final SqlQueryFactory factory;
+
+  /* pp */ TeradataLogsJdbcTask(
+      @Nonnull String targetPath, SharedState state, SqlQueryFactory factory) {
+    super(targetPath);
+    this.state = Preconditions.checkNotNull(state, "SharedState was null.");
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataLogsJdbcTask.java:34]: Parameters are by default not null. I would suggest to remove the `checkNotNull` call (similarly as you already don't check whether `factory` parameter is not null).

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactoryBuilder.java [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+        isOldVersion || !isAssessment
+            ? Collections.emptyList()
+            : Arrays.asList("ST.QueryID", "ST.SQLRowNo");
+  }
+
+  private void selectExpressions() {
+    if (isLSql) expressions = EXPRESSIONS_LSQL_TBL;
+    else if (isOldVersion) expressions = EXPRESSIONS_LOG_TBL;
+    else {
+      String[] selections = isAssessment ? EXPRESSIONS_FOR_ASSESSMENT : EXPRESSIONS;
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactoryBuilder.java:263]: I would suggest to change the type of `EXPRESSIONS_FOR_ASSESSMENT` and `EXPRESSIONS` to `ImmutableList<String>`.

**Review by dawidxc — COMMENTED** [review]
Thank you very much for taking the time to improve the quality of the SQL generation code. The resulting code in `TeradataSqlQueryFactory` looks much better after the changes.

I would suggest to split the PR into smaller ones. Possible pieces of this PR:
1. cut the ties between Teradata14LogsConnector and TeradataLogsConnector
- copy TeradataLogsJdbcTask to Teradata14LogsConnector and rename it to Teradata14LogsJdbcTask
- make Teradata14LogsConnector extends AbstractTeradataConnector implements LogsConnector, TeradataLogsDumpFormat
2. extraction of TeradataAssessmentLogsJdbcTask and TeradataLogsJdbcTask into separate java files
3. introduction of TeradataSqlQueryFactory
4. extraction of the `isValid` method from TeradataLogsJdbcTask to a separate class
5. usage of TeradataSqlQueryFactory in TeradataLogsJdbcTask and removal of TeradataAssessmentLogsJdbcTask
6. possibly some formatting-only PRs, if you need to format a single file or a few files (although this is optional, only if you want; I would encourage to perform the changes in the PRs above without reformatting the whole files for now, we should apply formatting to the whole codebase later)

File: dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactoryTest.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+  @Test
+  public void testGetSql() {
+    //  Arrange
+    ZonedDateTime endDate = ZonedDateTime.parse("2023-03-09T00:00:00Z");
+    ZonedDateTime startDate = endDate.minusDays(5).truncatedTo(ChronoUnit.DAYS);
+    ZonedInterval interval = new ZonedInterval(startDate, endDate);
+    TeradataSqlQueryFactory factory =
+        new TeradataSqlQueryFactory(
+            interval,
+            "LT",
```
Comment (frsv) [inline dumper/app/src/test/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactoryTest.java:23]: nit: I would suggest leaving a comment before string parameters for cases where it is not obvious what the parameter stands for. The comment will be equal to the parameter name, e.g. `/* logTable= */ "LT",`. As an alternative, you can extract them to variables or give more obvious names, such as `testLogTable`

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactoryBuilder.java [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+    return this;
+  }
+
+  private TeradataSqlQueryFactoryBuilder(boolean isLSql) {
+    this.isLSql = isLSql;
+  }
+
+  private void parse(ConnectorArguments arguments) throws MetadataDumperUsageException {
+    isAssessment = arguments.isAssessment();
+    isOldVersion = arguments.getConnectorName().startsWith("teradata14");
```
Comment (frsv) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/teradata/TeradataSqlQueryFactoryBuilder.java:217]: nit: Consider using other name for this variable, e.g. `isTeradata14Version`, as current can be misinterpreted as the old version of the dumper.
